### ISSUE TITLE
Allow public fields for immutable classes

### ISSUE THREAD

**Issue by KTannenberg** [checkstyle/checkstyle#61]
Affects: VisibilityModifierCheck

Rationale: Forcing all fields of class to have private modified by default is good in most cases, but it some cases it drawbacks in too much boilerplate get/set code. One of such cases are immutable classes.

Requested: 
* Allow immutable fields to be public. 
* Field is considered to be immutable if it have final modifier and either a primitive type or instance of class known to be immutable (such as String, ImmutableCollection from Guava and etc).
* Also it is would be nice to check, that class itself is also final.
* Classes known to be immutable should be listed by their fully qualified name

Example of immutable class that becomes much more concise with public fields
```java
public final class ReportItem {

	public final ImmutableSet<String> includes;
	public final ImmutableSet<String> excludes;
	public final String notes;
	public final BigDecimal value;

	public ReportItem(Collection<String> includes, Collection<String> excludes, BigDecimal value, String notes) {
		this.includes = ImmutableSet.copyOf(includes);
		this.excludes = ImmutableSet.copyOf(excludes);
		this.value = value;
		this.notes = notes;
	}
```

**Comment by asarazan** [comment]
+1 to suggestion by @jdalbey

**Comment by romani** [comment]
@KTannenberg ,

> Also it is would be nice to check, that class itself is also final

Please let me know why that validation is required ? As I can see final field could not be initialized in child classes, so not need to check "final" on Class. Example:
```
    class Base {
        final int field;
        Base () {
            field = 8; 
        }
        Base (int value) {
            field = value;
        }
    }
    
    class My extends Base {
        My (int value2) {
            field = value2; /// compile error - "The final field AutomaticBean.C.field cannot be assigned"
        }
    }
```

please provide code example to prove your idea.

**Comment by romani** [comment]
resolved, will be in 6.4

**Comment by romani** [comment]
> Please let me know why that validation is required ? As I can see final field could not be initialized in child classes, so not need to check "final" on Class

from private conversation with KTannenberg :

Making class and all its members final closely resembles Case classes from Scala (but with inheritance being disabled).

However if you make such classes non-final and thus allow inheritance from them (while still designing class as immutable), you can end up in situation where you have child class with mutable behaviour. There are couple of problems I see here:
  1. You would lose consistency in member access. As sole purpose of this check is to omit boilerplate get/set code in Base class, you would be forced to access parent's members with dot sythax and child's members with getter call
  2. You can put Child instances in Collection<Base> and get problems with hashCode()/equals(). And don't say me that noone would do such kind of crap... well maybe you and I wouldn't  :)

Best solution would be to use case classes from Scala, but there are also some cons:
  1. You would need to move to Scala
  2. It also have problems with Case class inheritance
